### DISCUSSION TITLE
SDK v2 Migration guide

### DISCUSSION THREAD

**Discussion by janossch** [Azure/azure-iot-sdk-java#1671] [Q&A] [answered]
Dear azure-iot-sdk-java developers,

Currently I'm trying to migrate my code to v2 libraries. And the very first thing I wanted to modify is the `ModuleClient.createEnvironment(...)` method.
As per the [migration guide](https://github.com/Azure/azure-iot-sdk-java/blob/main/SDK%20v2%20migration%20guide.md) suggesting the users can still depend on JNR unixsocket implementation. Nothing is simplier I've copied the provided [sample code](https://github.com/Azure/azure-iot-sdk-java/blob/main/iothub/device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java#L24).

1. However I don't really understand why there is a [System.arraycopy(...)](https://github.com/Azure/azure-iot-sdk-java/blob/main/iothub/device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java#L45) call is needed in the `read()` method?
`ByteBuffer.wrap(...) ` "modifications to the buffer will cause the array to be modified and vice versa" [source](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/ByteBuffer.html#wrap(byte%5B%5D)).

2. Since our project is based on Java17 do you think it is a good choice to use its built in [unix socket support](https://docs.oracle.com/en/java/javase/17/core/internet-protocol-and-unix-domain-sockets-nio-example.html)? 
... even if it is not among [supported JDK platforms](https://github.com/Azure/azure-iot-sdk-java/blob/main/supported_platforms.md#supported-jdk-versions) yet...

Thank you! I'm looking forward to your reply!

**Comment by timtay-microsoft** [comment, marked as answer]
> However I don't really understand why there is a [System.arraycopy(...)](https://github.com/Azure/azure-iot-sdk-java/blob/main/iothub/device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java#L45) call is needed in the read() method?
ByteBuffer.wrap(...)  "modifications to the buffer will cause the array to be modified and vice versa" [source](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/ByteBuffer.html#wrap(byte%5B%5D)).

This may be an unnecessary line in our sample code, and we'll take a look at it. You are right that it appears the wrapped buffer will
be updated without needing to call System.ArrayCopy(...) here as well.

> Since our project is based on Java17 do you think it is a good choice to use its built in [unix socket support](https://docs.oracle.com/en/java/javase/17/core/internet-protocol-and-unix-domain-sockets-nio-example.html)?
... even if it is not among [supported JDK platforms](https://github.com/Azure/azure-iot-sdk-java/blob/main/supported_platforms.md#supported-jdk-versions) yet...

Go for it! We use the JNR unix socket library in our samples largely because there was no native unix socket support in Java at the time.
